src/psyclone/parse/algorithm.py

```diff
@@ -0,0 +1,977 @@
+                       list(self._builtin_name_map.keys())))
+
+        from psyclone.parse.kernel import get_kernel_ast
+        modast = get_kernel_ast(module_name, self._alg_filename,
+                                self._kernel_path, self._line_length)
+        from psyclone.parse.kernel import KernelTypeFactory
+        return KernelCall(module_name,
+                          KernelTypeFactory(api=self._api).create(
+                              modast, name=kernel_name), args)
+
+    def update_arg_to_module_map(self, statement):
+        '''Takes a use statement and adds its contents to the internal
+        arg_name_to_module_name map. This map associates names
+        specified in the 'only' list with the corresponding use name.
+
+        :param statement: A use statement
+        :type statement: :py:class:`fparser.two.Fortran2003.Use_Stmt`
+        :raises InternalError: if the statement being passed is not an \
+        fparser use statement.
+
+        '''
+        # make sure statement is a use
+        if not isinstance(statement, Use_Stmt):
+            raise InternalError(
+                "algorithm.py:Parser:update_arg_to_module_map: Expected "
+                "a use statement but found instance of "
+                "'{0}'.".format(type(statement)))
+
+        use_name = str(statement.items[2])
+
+        # Extract only list. This can be removed when
+        # fparser#170 is implemented
+        if isinstance(statement.items[4], Only_List):
+            only_list = statement.items[4].items
+        else:
+            only_list = [statement.items[4]]
+
+        for item in only_list:
+            self._arg_name_to_module_name[str(item)] = use_name
+
+    def check_invoke_label(self, argument):
+        '''Takes the parse tree of an invoke argument containing an invoke
+        label. Raises an exception if this label has already been used
+        by another invoke in the same algorithm code. If all is well
+        it returns the label as a string.
+
+        :param argument: Parse tree of an invoke argument. This \
+        should contain a "name=xxx" argument.
+        :type argument: :py:class:`fparser.two.Actual_Arg_Spec`
+        :returns: the label as a string.
+        :rtype: str
+        :raises ParseError: if this label has already been used by
+        another invoke in this algorithm code.
+
+        '''
+        invoke_label = get_invoke_label(argument, self._alg_filename)
+        if invoke_label in self._unique_invoke_labels:
+            raise ParseError(
+                "Found multiple named invoke()'s with the same "
+                "label ('{0}') when parsing {1}".
+                format(invoke_label, self._alg_filename))
+        self._unique_invoke_labels.append(invoke_label)
+        return invoke_label
+
+# pylint: enable=too-many-arguments
+# pylint: enable=too-many-instance-attributes
+
+# Section 2: Support functions
+
+
+def parse_fp2(filename):
```

Comment:
Possibly this function should be moved to the 'parse.utils' module as it's not Algorithm specific. It could do with a better name than the one I came up with too but that's a secondary issue.

Reply:
OK, I've done this. I was keeping util functions in the files they are used if they are only used in one file but I agree this is a general function even if it is currently only used in one place. I've not renamed.